lib/preflight/preflighttest.js

```diff
@@ -0,0 +1,483 @@
+          if (remoteVideoTrackStats && remoteVideoTrackStats[0]) {
+            packetsLost += remoteVideoTrackStats[0].packetsLost;
+            packetsReceived += remoteAudioTrackStats[0].packetsReceived;
+          }
+          collectedStats.packetLoss.push(packetsReceived ? packetsLost * 100 / packetsReceived : 0);
+        }
+      });
+  }
+
+  function generatePreflightReport(collectedStats) {
+    testTiming.stop();
+    const selectedIceCandidatePairStats = collectedStats.selectedIceCandidatePairStats;
+    const isTurnRequired = selectedIceCandidatePairStats.localCandidate.candidateType === 'relay'
+    || selectedIceCandidatePairStats.remoteCandidate.candidateType === 'relay';
+
+    return {
+      roomSid: subscriberRoom.sid,
+      mediaRegion: subscriberRoom.mediaRegion,
+      signalingRegion: subscriberRoom.localParticipant.signalingRegion,
+      testTiming,
+      networkTiming: {
+        connect: connectTiming,
+        media: mediaTiming,
+      },
+      stats: {
+        jitter: makeStat(collectedStats.jitter),
+        rtt: makeStat(collectedStats.rtt),
+        outgoingBitrate: makeStat(collectedStats.outgoingBitrate),
+        incomingBitrate: makeStat(collectedStats.incomingBitrate),
+        packetLoss: makeStat(collectedStats.packetLoss),
+        networkQuality: makeStat(collectedStats.networkQuality),
+      },
+      selectedIceCandidatePairStats,
+      isTurnRequired,
+      iceCandidateStats: collectedStats.iceCandidateStats
+    };
+  }
+
+  function collectRTCStatsForDuration(duration, collectedStats = null) {
+    const startTime = Date.now();
+    const STAT_INTERVAL = 1000;
+    if (collectedStats === null) {
+      collectedStats = {
+        jitter: [],
+        rtt: [],
+        outgoingBitrate: [],
+        incomingBitrate: [],
+        packetLoss: [],
+        selectedIceCandidatePairStats: null,
+        iceCandidateStats: [],
+      };
+    }
+    return waitForSometime(STAT_INTERVAL).then(() => {
+      return collectRTCStats(collectedStats).then(() => {
+        const remainingDuration = duration - (Date.now() - startTime);
+        return (remainingDuration > 0) ? collectRTCStatsForDuration(remainingDuration, collectedStats) : collectedStats;
+      });
+    }).then(() => {
+      return collectIceCandidates().then(iceCandidates => {
+        collectedStats.iceCandidateStats = iceCandidates;
+        return collectedStats;
+      });
+    });
+  }
+
+  // @returns {Array<number>}
+  function collectNetworkQualityForDuration(duration) {
+    const networkQuality = [];
+    const localParticipant = subscriberRoom.localParticipant;
+    if (localParticipant.networkQualityLevel) {
+      networkQuality.push(localParticipant.networkQualityLevel);
+    }
+    const networkQualityCallback = () => networkQuality.push(localParticipant.networkQualityLevel);
+    localParticipant.addListener('networkQualityLevelChanged', networkQualityCallback);
+    return waitForSometime(duration).then(() => {
```

Comment:
Same here. Why do we need this? Will this affect time measurements?

Reply:
this ensures that function returns after collecting network quality score for `duration`. we watch for the  `networkQualityLevelChanged` during that time.